packages/loader/container-loader/src/container.ts

```diff
@@ -2436,11 +2453,92 @@ export class Container
 	}
 }
 
+/**
+ * Diagnostics information about a given connection instance (single attempt/clientId),
+ * traversing the four connection states (disconnected, establishingConnection, catchingUp, connected).
+ *
+ * Includes what the current connection state is, and what is happening within that state
+ *
+ * @beta
+ */
+export interface ConnectionDiagnostics {
+	/**
+	 * The client id assigned for this connection by the ordering service
+	 * Only set when currentState is "catchingUp" or "connected"
+	 */
+	clientId?: string;
+
+	//* This is hard to compute at the right time.  Maybe can follow example of ReconnectMode below?
+	/** read-only v. read/write connection */
+	connectionMode?: ConnectionMode;
+
+	//* TODO: Do we need additional modeling for the case when a connection attempt is cancelled?
+
+	/** The current (while active) or final (after disconnect) state of this connection */
+	state: keyof ConnectionDiagnostics["stateDetails"];
+
+	/** Details about the connection as it progressed through each state */
+	readonly stateDetails: {
+		/** Details about the connection while disconnected */
+		disconnected: {
+			readonly time: number;
+			readonly reason?: IConnectionStateChangeReason;
+
+			readonly autoReconnect: ReconnectMode;
+		};
+		/** Details about the connection while establishingConnection */
+		establishingConnection?: {
+			readonly time: number;
+			readonly reason?: IConnectionStateChangeReason;
+
+			steps: {
+				// 0 is current step
+				name: string; // Free-form, for telemetry only
+				type: "auth" | "socket.io" | "orderingService"; // unsure about this - part of public API, needs to be both stable and useful
+				time: number;
+				retryableError?: Error; // with errorType (maybe AnyDriverError)
+			}[];
+		};
+		/** Details about the connection while catchingUp */
+		catchingUp?: {
+			readonly time: number;
+			readonly reason?: IConnectionStateChangeReason;
+
+			checkpointSequenceNumber?: number;
+			initialProcessedSequenceNumber?: number;
+			currentProcessedSequenceNumber?: number;
+		};
+		/** Details about the connection while connected */
+		connected?: {
+			readonly time: number;
+			//* not used?
+			readonly reason?: IConnectionStateChangeReason;
+
+			opsSent?: number;
+			opsReceived?: number;
+		};
```

Comment:
I need to sleep on this, but my first impressions are the following:
1. This is "global" thing. As we add sub-steps in the system, we would need to keep this in sync with unrelated code. For example, how do we expose the fact that we are waiting for leave op for old connection? And how would we do it if that code moves across layers?
2. I'd rather model it as a set of events and let customers decide how they want to track it. We could provide them a helper that would listen for events and reconstruct structure like this one. But maybe someone else wants to do sampling and log every 10th second (while waiting for connection) on current state and that's it?

Reply:
Thanks for taking a look!

1. How complex of a state machine are we willing to build? The current system is not trivial but once I realized this I felt good: ConnectionStateHandler owns overall state machine, but other parts of system may own lifecycle of one state.

E.g. ConnectionManager owns EstablishingConnection state, and ConnectionStateHandler owns CatchingUp state.

I don't imagine we will completely deconstruct/distribute the state machine that there's not a global spot to hold this info.

2. I started with an events based concept but felt like it was too chatty or the events lacked context. But I'll revisit it. Comprehensive set of events would include "sub-step XYZ started" then "sub-step XYZ finished successfully" in order to enable someone to know what step is hung.

What I like about the "global" state always sitting there is that we can start to add events of interest, knowing that the app can always go check the full status in response to an event. There are no secrets. 

Btw - I put "global" in quotes because I don't think it's a bad kind of global - it's aligned with reality. The IContainer has a single-threaded sequence of connections or attempts.  If that changes, this would move the appropriate scope.